atc/db/pipeline_lifecycle.go

```diff
@@ -51,9 +53,20 @@ func (pl *pipelineLifecycle) ArchiveAbandonedPipelines() error {
 				// build that set the pipeline is not the most recent for the job.
 				// parent_build_id can be later than latest_completed_build_id if this
 				// gc query runs during a run of a build, specifically between the time
-				// of a completed set pipeline step and the build finishing
-				sq.Expr("p.parent_build_id < j.latest_completed_build_id"),
-			}}).
+				// of a completed set pipeline step and the build finishing. Also only
+				// take into account successful builds in order to know if this child
+				// pipeline had its set_pipeline step removed from parent job.
+				sq.And{
+					sq.Expr("p.parent_build_id < j.latest_completed_build_id"),
+					sq.Expr(`EXISTS (
```

Comment:
I still don't think this is good enough. For example, `set_pipeline` is wrapped under `on_failure`, then after a success run, child pipeline will be archived.

`p.parent_build_id < j.latest_completed_build_id` is used to handle a case where a job originally had `set_pipeline` then removed `set_pipeline` from the job, then all child pipelines set by this job would be archived. As we have seen that, this logic is hard to implement, it relates pipeline logic, there are a lot of cases.

I still cannot understand why many users reported auto archive bug after upgrade to 7.7. You changed "!=" to "<", the condition actually became to less aggressive.

So I think maybe we just remove `p.parent_build_id < j.latest_completed_build_id`. Only auto archive child pipeline when parent destroyed or archived.

Reply:
I agree this is a hard case to implement, there's many edge cases to it that are hard to completely flush out. I do think that this is an important feature to get down though because we want to be able to support child pipelines getting set and archived when it is no longer a step within the parent job. Our use case for this is our pr pipelines within our prod environment. https://ci.concourse-ci.org/teams/main/pipelines/reconfigure-pipelines/jobs/reconfigure-prs/builds/792

The case that you mentioned with `on_failure` is a very good point, I agree that will break this behaviour. I'm not sure I can think of a reason why someone would want to do a `set_pipeline` during an `on_failure` though because the purpose of `on_failure`, `on_error`, etc is for notifications. But we do not restrict a user from using it in this way. Maybe there is a better way to achieve the result we want without using this current method of comparing `parent_build_id` and `latest_completed_build_id`, I'll think about it some more because we do want the end result feature.

For the auto archive bug after upgrade, the PR https://github.com/concourse/concourse/pull/8137 never made it into 7.7.0.

Reply:
I merged this pr so that I can include it in the 7.7.1 patch release but I created an issue here https://github.com/concourse/concourse/issues/8201 for the concerns we have mentioned in this thread and for rethinking the logic for auto archiving. Once we get that fixed, we will get it into 7.8.0.